File: go/lesson-02-functions/src/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	showBasicFunctions()
	fmt.Println("---")
	showMultipleReturns()
	fmt.Println("---")
	showVariadic()
	fmt.Println("---")
	showFirstClass()
	fmt.Println("---")
	showClosures()
}

// Basic function — parameters of same type can share type annotation
func add(a, b int) int {
	return a + b
}

func greet(name string) {
	fmt.Printf("Hello, %s!\n", name)
}

func showBasicFunctions() {
	fmt.Println("=== Basic Functions ===")
	fmt.Println("3 + 4 =", add(3, 4))
	greet("Alice")
}

// Multiple return values — idiomatic Go
func divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, fmt.Errorf("cannot divide %.2f by zero", a)
	}
	return a / b, nil
}

// Named return values — useful for documenting what is returned
func circleMetrics(radius float64) (area, circumference float64) {
	area = math.Pi * radius * radius
	circumference = 2 * math.Pi * radius
	return // naked return — returns named values
}

func showMultipleReturns() {
	fmt.Println("=== Multiple Return Values ===")

	result, err := divide(10, 3)
	if err != nil {
		fmt.Println("Error:", err)
	} else {
		fmt.Printf("10 / 3 = %.4f\n", result)
	}

	_, err = divide(5, 0)
	if err != nil {
		fmt.Println("Error:", err)
	}

	area, circ := circleMetrics(5)
	fmt.Printf("Circle r=5: area=%.2f circumference=%.2f\n", area, circ)
}

// Variadic function — receives args as a slice
func sum(nums ...int) int {
	total := 0
	for _, n := range nums {
		total += n
	}
	return total
}

func max(first int, rest ...int) int {
	m := first
	for _, n := range rest {
		if n > m {
			m = n
		}
	}
	return m
}

func showVariadic() {
	fmt.Println("=== Variadic Functions ===")
	fmt.Println("sum(1,2,3):", sum(1, 2, 3))
	fmt.Println("sum(1..10):", sum(1, 2, 3, 4, 5, 6, 7, 8, 9, 10))

	nums := []int{4, 7, 2, 9, 1}
	fmt.Println("sum(slice...):", sum(nums...)) // spread slice

	fmt.Println("max(3,1,4,1,5,9,2,6):", max(3, 1, 4, 1, 5, 9, 2, 6))
}

// Higher-order function — accepts a function as argument
func apply(nums []int, fn func(int) int) []int {
	result := make([]int, len(nums))
	for i, n := range nums {
		result[i] = fn(n)
	}
	return result
}

func filter(nums []int, predicate func(int) bool) []int {
	var result []int
	for _, n := range nums {
		if predicate(n) {
			result = append(result, n)
		}
	}
	return result
}

func showFirstClass() {
	fmt.Println("=== First-Class Functions ===")

	// Function assigned to a variable
	double := func(n int) int { return n * 2 }
	square := func(n int) int { return n * n }

	nums := []int{1, 2, 3, 4, 5}
	fmt.Println("original:", nums)
	fmt.Println("doubled: ", apply(nums, double))
	fmt.Println("squared: ", apply(nums, square))

	evens := filter(nums, func(n int) bool { return n%2 == 0 })
	fmt.Println("evens:   ", evens)
}

// Closure — captures variables from the enclosing scope
func makeCounter(start int) func() int {
	n := start
	return func() int {
		n++
		return n
	}
}

func makeAdder(x int) func(int) int {
	return func(y int) int {
		return x + y // captures x
	}
}

func makeMultiplier(factor int) func(int) int {
	return func(n int) int { return n * factor }
}

func showClosures() {
	fmt.Println("=== Closures ===")

	// Each counter has its own independent state
	c1 := makeCounter(0)
	c2 := makeCounter(10)
	fmt.Printf("c1: %d %d %d\n", c1(), c1(), c1()) // 1 2 3
	fmt.Printf("c2: %d %d %d\n", c2(), c2(), c2()) // 11 12 13

	add5 := makeAdder(5)
	fmt.Println("add5(3):", add5(3))   // 8
	fmt.Println("add5(10):", add5(10)) // 15

	triple := makeMultiplier(3)
	nums := []int{1, 2, 3, 4, 5}
	fmt.Println("tripled:", apply(nums, triple))
}

File: nodejs/lesson-02-functions/src/main.js
'use strict';

showBasicFunctions();
console.log('---');
showArrowFunctions();
console.log('---');
showRestSpread();
console.log('---');
showClosures();
console.log('---');
showHigherOrder();

// ─── Basic Functions ──────────────────────────────────────────────────────────

function showBasicFunctions() {
  console.log('=== Basic Functions ===');

  // Declaration — hoisted (can call before definition in same scope)
  console.log('add(3, 4):', add(3, 4));

  function add(a, b) { return a + b; }

  // Expression — not hoisted
  const multiply = function(a, b) { return a * b; };
  console.log('multiply(3, 4):', multiply(3, 4));

  // Default parameters
  function greet(name = 'World', greeting = 'Hello') {
    return `${greeting}, ${name}!`;
  }
  console.log(greet());              // Hello, World!
  console.log(greet('Alice'));       // Hello, Alice!
  console.log(greet('Bob', 'Hi'));   // Hi, Bob!
}

// ─── Arrow Functions ──────────────────────────────────────────────────────────

function showArrowFunctions() {
  console.log('=== Arrow Functions ===');

  // Concise body — implicit return
  const square = n => n * n;
  const add = (a, b) => a + b;

  // Block body — explicit return needed
  const clamp = (n, min, max) => {
    if (n < min) return min;
    if (n > max) return max;
    return n;
  };

  console.log('square(5):', square(5));
  console.log('add(3, 4):', add(3, 4));
  console.log('clamp(15, 0, 10):', clamp(15, 0, 10));

  // Arrow functions have no own `this` — they capture it from enclosing scope
  // This is why they're preferred as callbacks and class property methods
  const timer = {
    id: 'T1',
    // Arrow function: `this` refers to `timer` object
    start() {
      const fn = () => this.id; // `this` = timer
      return fn();
    },
  };
  console.log('timer.start():', timer.start()); // T1

  // Immediately invoked arrow
  const result = ((x, y) => x + y)(3, 4);
  console.log('IIFE result:', result); // 7
}

// ─── Rest & Spread ────────────────────────────────────────────────────────────

function showRestSpread() {
  console.log('=== Rest & Spread ===');

  // Rest — collect remaining args into an array
  function sum(...nums) {
    return nums.reduce((total, n) => total + n, 0);
  }
  console.log('sum(1,2,3,4,5):', sum(1, 2, 3, 4, 5)); // 15

  function first(head, ...tail) {
    return { head, tail };
  }
  console.log('first(1,2,3):', first(1, 2, 3)); // { head: 1, tail: [2, 3] }

  // Spread — expand array/object
  const nums = [3, 1, 4, 1, 5];
  console.log('Math.max(...nums):', Math.max(...nums)); // 5
  console.log('sum(...nums):', sum(...nums)); // 14

  const arr1 = [1, 2, 3];
  const arr2 = [4, 5, 6];
  console.log('combined:', [...arr1, ...arr2]); // [1,2,3,4,5,6]

  const defaults = { theme: 'light', lang: 'en', timeout: 3000 };
  const overrides = { lang: 'vi', timeout: 5000 };
  const config = { ...defaults, ...overrides };
  console.log('config:', config);

  // Spread to clone (shallow)
  const original = { a: 1, b: { c: 2 } };
  const clone = { ...original };
  clone.a = 99;
  console.log('original.a:', original.a); // 1 (unchanged)
}

// ─── Closures ─────────────────────────────────────────────────────────────────

function showClosures() {
  console.log('=== Closures ===');

  // Counter — each call to makeCounter creates independent state
  function makeCounter(start = 0) {
    let n = start;
    return {
      increment: () => ++n,
      decrement: () => --n,
      reset: () => { n = start; },
      value: () => n,
    };
  }

  const c1 = makeCounter(0);
  const c2 = makeCounter(10);
  c1.increment(); c1.increment(); c1.increment();
  c2.decrement();
  console.log(`c1=${c1.value()} c2=${c2.value()}`); // c1=3 c2=9

  // Memoize — cache computed results
  function memoize(fn) {
    const cache = new Map();
    return function(...args) {
      const key = JSON.stringify(args);
      if (cache.has(key)) {
        console.log(`  cache hit for ${key}`);
        return cache.get(key);
      }
      const result = fn(...args);
      cache.set(key, result);
      return result;
    };
  }

  const slowSquare = memoize(n => {
    // simulate expensive computation
    return n * n;
  });
  console.log(slowSquare(5)); // computed
  console.log(slowSquare(5)); // cache hit
  console.log(slowSquare(6)); // computed
}

// ─── Higher-Order Functions ───────────────────────────────────────────────────

function showHigherOrder() {
  console.log('=== Higher-Order Functions ===');

  // map, filter, reduce — the core trio
  const nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
  const evens = nums.filter(n => n % 2 === 0);
  const doubled = evens.map(n => n * 2);
  const total = doubled.reduce((sum, n) => sum + n, 0);
  console.log('evens:', evens);
  console.log('doubled evens:', doubled);
  console.log('sum of doubled evens:', total); // 60

  // Chain pipeline
  const result = nums
    .filter(n => n % 2 === 0)
    .map(n => n * 2)
    .reduce((sum, n) => sum + n, 0);
  console.log('chained:', result); // 60

  // Compose — right-to-left function pipeline
  const compose = (...fns) => x => fns.reduceRight((v, fn) => fn(v), x);
  const pipe = (...fns) => x => fns.reduce((v, fn) => fn(v), x);

  const double = n => n * 2;
  const addOne = n => n + 1;
  const square = n => n * n;

  const transform = pipe(double, addOne, square);
  console.log('pipe(double, addOne, square)(3):', transform(3)); // (3*2+1)^2 = 49

  // Partial application — fix some arguments
  function partial(fn, ...presetArgs) {
    return (...laterArgs) => fn(...presetArgs, ...laterArgs);
  }

  const multiply = (a, b) => a * b;
  const triple = partial(multiply, 3);
  console.log('triple(5):', triple(5));   // 15
  console.log('triple(10):', triple(10)); // 30
}
